src/mame/sinclair/specnext.cpp

```diff
@@ -0,0 +1,2757 @@
+
+
+namespace {
+
+class specnext_state : public spectrum_128_state
+{
+public:
+	specnext_state(const machine_config &mconfig, device_type type, const char *tag)
+		: spectrum_128_state(mconfig, type, tag)
+		, m_maincpu(*this, "maincpu")
+		, m_bank_boot_rom(*this, "bootrom")
+		, m_bank_ram(*this, "bank_ram%u", 0U)
+		, m_view0(*this, "mem_view0")
+		, m_view1(*this, "mem_view1")
+		, m_view2(*this, "mem_view2")
+		, m_view3(*this, "mem_view3")
+		, m_view4(*this, "mem_view4")
+		, m_view5(*this, "mem_view5")
+		, m_view6(*this, "mem_view6")
+		, m_view7(*this, "mem_view7")
+		, m_ctc(*this, "ctc")
+		, m_z80dma(*this, "z80dma")
+		, m_eeprom(*this, "eeprom")
+		, m_sdcard(*this, "sdcard")
+		, m_centronics(*this, "centronics")
+		, m_palette(*this, "palette")
+		, m_regs_map(*this, "regs_map")
+		, m_divmmc(*this, "divmmc")
+		, m_ula(*this, "ula")
+		, m_tiles(*this, "tiles")
+		, m_layer2(*this, "layer2")
+		, m_sprites(*this, "sprites")
+		, m_io_issue(*this, "ISSUE")
+		, m_io_mouse(*this, "mouse_input%u", 1U)
+	{}
+
+	void tbblue(machine_config &config);
+
+protected:
+	virtual void machine_start() override;
+	virtual void machine_reset() override;
+	virtual void video_start() override;
+	virtual void spectrum_128_update_memory() override {};
+
+	u8 on_fetch_divmmc(offs_t offset);
+	void map_fetch(address_map &map);
+	void map_mem(address_map &map);
+	void map_io(address_map &map);
+	void map_regs(address_map &map);
+	u8 reg_r(offs_t reg);
+	void reg_w(offs_t reg, u8 data);
+	void mmu_w(offs_t bank, u8 data);
+	void cpu_speed_w(u8 data);
+	u8 spi_data_r();
+	void spi_data_w(u8 data);
+	void spi_miso_w(u8 data);
+	void palette_val_w(u16 nr_palette_value);
+	void ula_palette_w(u8 data);
+
+	void mem_update();
+	void bank_update(u8 bank);
+	void memory_change(u16 port, u8 data);
+	u16 get_layer2_active_page(u8 bank);
+	virtual u32 screen_update_spectrum(screen_device &screen, bitmap_ind16 &bitmap, const rectangle &cliprect) override;
+	virtual void spectrum_update_screen(screen_device &screen, bitmap_ind16 &bitmap, const rectangle &cliprect) override {}; // ULA screen just border
+
+	required_device<z80n_device> m_maincpu;
+
+private:
+	static const u8 MACHINE_TBBLUE = 0x08;
+	static const u8 MACHINE_NEXT = 0x0a;
+	static const u8 MACHINE_NEXT_AB = 0xfa; // Anti Brick (reset disabled, bootrom)
+	static const u8 G_VERSION = 0x32; // 3.02
+	static const u8 G_SUB_VERSION = 0x01;
+	static const u8 G_VIDEO_INC = 0b11;
+	static const u16 UTM_FALLBACK_PEN = 0x800;
+
+	TIMER_CALLBACK_MEMBER(spi_clock);
+
+	u8 g_machine_id() { return m_io_issue->read() ? MACHINE_NEXT : MACHINE_TBBLUE; }
+	u8 g_board_issue() { return m_io_issue->read(); }
+	constexpr bool machine_type_48() { return m_regs.nr_03_machine_type == 0 || m_regs.nr_03_machine_type == 1; }
+	constexpr bool machine_type_128() { return m_regs.nr_03_machine_type == 2 || m_regs.nr_03_machine_type == 4; }
+	constexpr bool machine_type_p3() { return !machine_type_48() && !machine_type_128(); }
```

Comment:
Do `constexpr` member functions make sense at all for a device class?  You can’t have have compile time constant instances of it.  Wouldn’t it make more sense to just make them `const` qualified?